usr/lib/cca_stdll/cca_specific.c

```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+    case CKA_SUBPRIME_BITS: typestr = "CKA_SUBPRIME_BITS"; break;
+    case CKA_VALUE_BITS: typestr = "CKA_VALUE_BITS"; break;
+    case CKA_VALUE_LEN: typestr = "CKA_VALUE_LEN"; break;
+    case CKA_EXTRACTABLE: typestr = "CKA_EXTRACTABLE"; break;
+    case CKA_LOCAL: typestr = "CKA_LOCAL"; break;
+    case CKA_NEVER_EXTRACTABLE: typestr = "CKA_NEVER_EXTRACTABLE"; break;
+    case CKA_ALWAYS_SENSITIVE: typestr = "CKA_ALWAYS_SENSITIVE"; break;
+    case CKA_KEY_GEN_MECHANISM: typestr = "CKA_KEY_GEN_MECHANISM"; break;
+    case CKA_MODIFIABLE: typestr = "CKA_MODIFIABLE"; break;
+    case CKA_EC_PARAMS: typestr = "CKA_EC_PARAMS"; break;
+    case CKA_EC_POINT: typestr = "CKA_EC_POINT"; break;
+    case CKA_ALWAYS_AUTHENTICATE: typestr = "CKA_ALWAYS_AUTHENTICATE"; break;
+    case CKA_IBM_OPAQUE: typestr = "CKA_IBM_OPAQUE"; break;
+    default:
+        sprintf(tbuf, "unknown type 0x%08lx", a->type);
+        typestr = tbuf;
+        break;
+    }
+
+    switch (a->ulValueLen) {
+    case 0:
+        TRACE_DEBUG("  %s: len=0 pValue=%p\n", typestr, a->pValue);
+        break;
+    case 1:
+        TRACE_DEBUG("  %s: len=%lu value=0x%02hhx\n",
+                    typestr, a->ulValueLen, *((uint8_t *)(a->pValue)));
+        break;
+    case 2:
+        TRACE_DEBUG("  %s: len=%lu value=0x%04hx\n",
+                    typestr, a->ulValueLen, *((uint16_t *)(a->pValue)));
+        break;
+    case 4:
+        TRACE_DEBUG("  %s: len=%lu value=0x%08x\n",
+                    typestr, a->ulValueLen, *((uint32_t *)(a->pValue)));
+        break;
+    case 8:
+        TRACE_DEBUG("  %s: len=%lu value=0x%016lx\n",
+                    typestr, a->ulValueLen, *((uint64_t *)(a->pValue)));
+        break;
+    default:
+        TRACE_DEBUG("  %s: len=%lu value:\n", typestr, a->ulValueLen);
+        dump_mem("  ", a->pValue, a->ulValueLen);
+        break;
+    }
+}
+
+/* Debug function: dump list of attribues from a template via TRACE_DEBUG */
+static void dump_template(TEMPLATE *tmpl)
+{
+    DL_NODE *node = NULL;
+    CK_ATTRIBUTE *a = NULL;
+
+    node = tmpl->attribute_list;
+    while (node) {
+        a = (CK_ATTRIBUTE *) node->data;
+        dump_attr(a);
+        node = node->next;
+    }
+}
+#endif
+
+/* cca token type enum, used with analyse_cca_key_token() */
+enum cca_token_type {
+    sec_des_data_key,
+    sec_aes_data_key,
+    sec_aes_cipher_key,
+    sec_hmac_key,
+    sec_rsa_priv_key,
+    sec_rsa_publ_key,
+    sec_ecc_priv_key,
+    sec_ecc_publ_key
+};
+
+/*
+ * Helper function: Analyse given CCA token.
+ * returns TRUE and keytype and keybitsize if token is known and seems
+ * to be valid (only basic checks are done), otherwise FALSE.
+ */
+static CK_BBOOL analyse_cca_key_token(const CK_BYTE *t, CK_ULONG tlen,
+                                      enum cca_token_type *keytype,
+                                      unsigned int *keybitsize)
+{
+    if (t[0] == 0x01 && (t[4] == 0x00 || t[4] == 0x01)) {
```

Comment:
I usually like struct to overlay over the buffer better than hard coded offsets, but since the CCA token uses the offset-way already, I am fine with this, too.

Reply:
I have the structs - but as most of the cca specific code deals with some kind of offsets and would need a total rework to adapt to structs, I've chosen to use these simple offset values.